### DISCUSSION TITLE
Any XState util function to sync actor refs in the context?

### DISCUSSION THREAD

**Discussion by BrunoQuaresma** [statelyai/xstate#3406] [Q&A]
XState folks is there any chance of XState having a util function to "sync" actors in a context? I mean, let's there is a list and each item on the list is an actor. The list machine is pooling updating the list data every 5 seconds.

After getting the new data some items can be there, or not. Some items maybe are deleted. The first thing that I thought was to re-spawn those actors after each update but it can mess up the state of each item if they are executing some action.

So, right now, I'm iterating through the data and adding new items if they are new, deleting the refs that are not present anymore, and updating the ones that are still present but it looks kinda "janky" for me to have to do that. 

Here is the code that I came up with:
```typescript
updateWorkspaceRefs: assign({
  workspaceRefs: (context, event) => {
    let workspaceRefs = context.workspaceRefs

    if (!workspaceRefs) {
      throw new Error("No workspaces loaded.")
    }

    // Update the existent refs or create the new ones
    for (const data of event.data) {
      const ref = workspaceRefs.find((ref) => ref.id === data.id)

      if (!ref) {
        workspaceRefs.push(spawn(workspaceItemMachine.withContext({ data }), data.id))
      } else {
        ref.send({ type: "UPDATE_DATA", data })
      }
    }

    // Remove refs that were deleted
    for (const ref of workspaceRefs) {
      const refData = event.data.find((workspaceData) => workspaceData.id === ref.id)

      // If there is no refData, it is because the ref was deleted
      if (!refData) {
        // Stop the actor before remove it from the array
        if (ref.stop) {
          ref.stop()
        }

        // Remove ref from the array
        workspaceRefs = workspaceRefs.filter((oldRef) => oldRef.id === ref.id)
      }
    }

    return workspaceRefs
  },
})
```

Thanks in advance!

**Comment by Andarist** [comment]
Hm, so you want to synchronize a list of actors to a list of data closely related to that list of actors? It feels like [receptionist](https://github.com/statelyai/rfcs/pull/5) could help you in the future. We were also thinking about "actor groups" within children - with those you could just send an event to a group and XState would loop through the existing actors of that group. Both ideas are somewhat similar - the difference is that receptionist is global~ but those actor groups could be scoped per parent.

Just note that I would advise you to use `pure` here - side-effects within `assign` are not encouraged.

**Reply by BrunoQuaresma** [reply]
I took a look at the RFC but I'm also not sure how the receptionist could help with this kinda sync. 🤔

**Comment by BrunoQuaresma** [comment]
Ah thanks. Why is this considered a side effect?

**Reply by Andarist** [reply]
Both `.stop()` and `.send()` used like that are actually doing stuff. "Resolving" `assign` becomes impure cause of that.

**Reply by BrunoQuaresma** [reply]
About `pure()`, I need to assign the refs to the context, because if some of them are deleted I want to remove the item from the array, and I see reading the docs that the `pure()` expects a different type of return.
```typescript
const sendToAllSampleActors = pure((context, event) => {
  return context.sampleActors.map((sampleActor) => {
    return send('SOME_EVENT', { to: sampleActor });
  });
});
```
